Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp

```diff
@@ -32,6 +32,66 @@
 
 using namespace amrex;
 
+SpectralFieldIndex::SpectralFieldIndex (const bool update_with_rho,
+                                        const bool time_averaging,
+                                        const bool J_linear_in_time,
+                                        const bool dive_cleaning,
+                                        const bool divb_cleaning,
+                                        const bool pml)
+{
+    // TODO Use these to allocate rho_old, rho_new, F, and G only when needed
+    amrex::ignore_unused(update_with_rho);
+
+    int c = 0;
+
+    if (pml == false)
+    {
+        Ex = c++; Ey = c++; Ez = c++;
+        Bx = c++; By = c++; Bz = c++;
+        Jx = c++; Jy = c++; Jz = c++;
+
+        // TODO Allocate rho_old and rho_new only when needed
```

Comment:
Does this simply consist in:
```
if (update_with_rho) {
    rho_old = c++; rho_new = c++;
}
```
or is it more complex than this?

Reply:
It is a bit more complicated than this, because we would have to make sure that we never access the corresponding data elsewhere, unless we need to. For instance, in https://github.com/ECP-WarpX/WarpX/blob/e94122ce9a9d61c7d22ea25b593cbf04d0f5bf8b/Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithm.cpp#L186-L187
we would need to always declare `rho_old` and `rho_new`, but actually access them only if necessary (e.g. when `update_with_rho = true` etc.), which is not the case currently.

Since this will change a bit our data storage, and is not just refactoring the code by using a different design for the spectral indices, I thought I could do that more consistently in a separate PR, which would address the optimization of the data stored in spectral space, for the different runtime variants that are not covered yet. The same holds for the field data F and G mentioned in your other comment below.

Would that be OK if these data storage optimizations are taken care of in a different PR?